### PR TITLE
Added verification of ascii inside unicode conditional

### PR DESCRIPTION
**Pull request by mday299**
Added verification of ascii inside unicode conditional

### CODE REVIEW

File: tools/mavtomfile.py [hunk 1]
```diff
@@ -74,6 +74,8 @@ def process_tlog(filename):
         f.write("%s.data(%u,:) = [%f" % (mtype, type_counters[mtype], m._timestamp))
         for field in m._fieldnames:
             val = getattr(m, field)
+            if isinstance(val,unicode):
```
Comment (peterbarker) [inline tools/mavtomfile.py:77]:
```suggestion
            if isinstance(val, unicode):
```

Comment (QuPsi) [reply tools/mavtomfile.py:77]: To be honest, I'm not really sure right now. The issue seems to be from how certain .BIN log files are created. I'll be looking into this to offer more information when I can.

Comment (QuPsi) [reply tools/mavtomfile.py:77]: Hi, my apologies for the delay, but I have been exploring this. I ran a simulation and generated a .BIN file. I then used a debugger to take note which values were being stored as unicode. 

The snippet below is from the txt output of the same .BIN file. This output was generated using MAVlopdump.py.

2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 45, Label : }
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 63, Label : UNKNOWN}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 65, Label : A}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 100, Label : deg}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 98, Label : B}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 107, Label : deg/s}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 68, Label : deglatitude}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 101, Label : deg/s/s}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 69, Label : rad/s}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 71, Label : Gauss}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 104, Label : degheading}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 105, Label : A.s}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 74, Label : W.s}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 76, Label : rad/s/s}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 109, Label : m}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 110, Label : m/s}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 111, Label : m/s/s}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 79, Label : degC}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 37, Label : %}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 83, Label : satellites}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 115, Label : s}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 113, Label : rpm}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 114, Label : rad}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 85, Label : deglongitude}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 117, Label : ppm}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 118, Label : V}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 80, Label : Pa}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 119, Label : Ohm}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 89, Label : us}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 122, Label : Hz}
2020-08-04 14:37:54.50: UNIT {TimeUS : 1864913736, Id : 35, Label : instance}

All of the values for "Label" are stored as unicode. There are others but those led me to LogStructure.h. There were other strings labeled as unicode that were also found in LogStructure.h (e.g. "s---SmDUmnhn-"). So it's clear that these strings are coming from LogStructure.h however, WHY they're being stored as unicode in the .BIN files I have not been able to figure out. I will continue to explore, but this is what I've found so far.

**Comment by peterbarker** [comment]
On Wed, 5 Aug 2020, QuPsi wrote:

> All of the values for "Label" are stored as unicode. There are others but those led me to LogStructure.h. There were other strings labeled as unicode that
> were also found in LogStructure.h (e.g. "s---SmDUmnhn-"). So it's clear that these strings are coming from LogStructure.h however, WHY they're being
> stored as unicode in the .BIN files I have not been able to figure out. I will continue to explore, but this is what I've found so far.

That's a function of Python string handling rather than anything else. 
Those strings are pure ASCII - to my knowledge we don't include any byte 
in those strings with the high bit set - and I'm actually not sure what 
the mavlink spec says on the matter (my guess is that it doesn't cover 
it...).

When you say, "labeled as unicode" - who was doing the labelling?

**Comment by QuPsi** [comment]
This is what the debugger reports when running the mavtomfile script and checking each individual value. Here's a list of what python sees as unicode:

u''
u'AK09916: Reset Failed\r\n'
u'Ground start'
u'Beginning INS calibration. Do not move plane'
u'AHRS: resetting...'
u'Calibrating barometer'
u'Barometer 1 calibration complete'
u'Barometer 2 calibration complete'
u'Airspeed calibration started'
u'Ground start complete'
u'UNKNOWN'
u'A'
u'deg'
u'B'
u'deg/s'
u'deglatitude'
u'deg/s/s'
u'rad/s'
u'Gauss'
u'degheading'
u'A.s'
u'W.s'
u'rad/s/s'


Pretty much all the values from my previous comment. There's more of course, but I think this is enough to report. Without the check and explicit conversion to ascii, the script will fail.

**Comment by peterbarker** [comment]
@QuPsi Can you give me steps to reproduce the problem, please?

**Comment by QuPsi** [comment]
To create the problem, I simply ran a SITL simulation with logging enabled. Then I pulled the log from the log folder in ardupilot and used mavtomfile.py to convert it to matlab format. Without the modification, this is the error mavtomfile.py throws:

Traceback (most recent call last):
  File ".../ardupilot/modules/mavlink/pymavlink/tools/mavtomfile.py", line 101, in <module>
    process_tlog(filename)
  File ".../ardupilot/modules/mavlink/pymavlink/tools/mavtomfile.py", line 82, in process_tlog
    f.write(",%.20g" % val)
TypeError: float argument required, not unicode

I've tested a number of log files that I have access to and have received the same error in all of them.

**Comment by peterbarker** [comment]
@QuPsi thanks for the instructions to reproduce.

I think the correct fix here might be to treat type-unicode just like we do type-str here:
https://github.com/ArduPilot/pymavlink/blob/master/tools/mavtomfile.py#L65
and here:
https://github.com/ArduPilot/pymavlink/blob/master/tools/mavtomfile.py#L77

**Comment by QuPsi** [comment]
Updated fix here.

#447

**Comment by peterbarker** [comment]
This is being handled in #447

**Pull request closed without merging**